### PR TITLE
Improve testing coverage of internal/app/singularity

### PR DESCRIPTION
**Pull request by gvallee**
**Description of the Pull Request (PR):**

Improve testing coverage of internal/app/singularity, focusing on `RemoteAdd()` and `RemoteRemove()`

**This fixes or addresses the following GitHub issues:**

- Addresses #3315 

**Before submitting a PR, make sure you have done the following:**

- Read the [Guidelines for Contributing](https://github.com/sylabs/singularity/blob/master/CONTRIBUTING.md), and this PR conforms to the stated requirements.
- Added changes to the [CHANGELOG](https://github.com/sylabs/singularity/blob/master/CHANGELOG.md) if necessary according to the [Contribution Guidelines](https://github.com/sylabs/singularity/blob/master/CONTRIBUTING.md)
- Added tests to validate this PR and tested this PR locally with a `make testall`
- Based this PR against the appropriate branch according to the [Contribution Guidelines](https://github.com/sylabs/singularity/blob/master/CONTRIBUTING.md)
- Added myself as a contributor to the [Contributors File](https://github.com/sylabs/singularity/blob/master/CONTRIBUTORS.md)

Attn: @singularity-maintainers

### CODE REVIEW

**Comment by gvallee** [comment]
I would like to have feedback about the cases that I am testing in the context of `RemoteAdd()` and `RemoteRemove()`. Specifically, I would like to have feedback on whether some of the corner cases that currently pass should in fact fail and whether some error paths should success. In other words, I would like to clarify the intent.

File: internal/app/singularity/remote_add_test.go [hunk 1]
```diff
@@ -0,0 +1,300 @@
...
+			// RemoveRemove().
+			name:       "9: valid config file; valid remote name; valid URI; local",
+			cfgfile:    validCfgFile,
+			remoteName: "cloud_testing",
+			uri:        "cloud.random.io",
+			global:     false,
+			shallPass:  true,
+		},
+		{
+			name:       "10: valid but dummy file; empty remote name; empty URI; local",
```
Comment (ikaneshiro) [inline internal/app/singularity/remote_add_test.go]: I think this is a duplicate of 2?

File: internal/app/singularity/remote_add_test.go [hunk 2]
```diff
@@ -0,0 +1,300 @@
...
+	}
+	cfg := aDummyStruct{
+		NoneSenseRemote: "toto",
+	}
+
+	yaml, err := yaml.Marshal(cfg)
+	if err != nil {
+		f.Close()
+		os.Remove(path)
+		t.Fatalf("cannot marshal YANK: %s\n", err)
```
Comment (ikaneshiro) [inline internal/app/singularity/remote_add_test.go]: Typo: `YANK`->`YAML`

**Review by ikaneshiro — CHANGES_REQUESTED** [review]
I think several of these corner cases should indeed fail. I think the `remove` behavior is ok but here are the cases I think `add` should catch and fail:
- Empty remote name
- Empty remote URI

With respect to `invalidCfgFile`. I think this should always cause a failure but this needs to be fixed in `ReadFrom()` in the `remote` package by having it use to use `UnmarshallStrict()`.

**Comment by gvallee** [comment]
Thanks for the feedback, I will address of your comments and update the PR.

**Comment by gvallee** [comment]
I create issue #3398 to track and fix the problem with ReadFrom().

File: internal/app/singularity/remote_add_test.go (hunk 1 above)
Comment (gvallee) [reply internal/app/singularity/remote_add_test.go]: I made a pass over all the cases, removed the duplicates and added the missing cases.

File: internal/app/singularity/remote_add_test.go (hunk 2 above)
Comment (gvallee) [reply internal/app/singularity/remote_add_test.go]: fixed

Comment (ikaneshiro) [reply internal/app/singularity/remote_add_test.go]: Looks like the typo is still there

File: internal/app/singularity/remote_add.go [hunk 3]
```diff
@@ -10,12 +10,21 @@ import (
...
+	"strings"
 
 	"github.com/sylabs/singularity/internal/pkg/remote"
 )
 
 // RemoteAdd adds remote to configuration
 func RemoteAdd(configFile, name, uri string, global bool) (err error) {
+	// Explicit handling of corner cases: name and uri must be valid strings
+	if strings.TrimSpace(name) == "" {
+		return fmt.Errorf("invalid name: %s", name)
```
Comment (ikaneshiro) [inline internal/app/singularity/remote_add.go]: Printing an error with an empty string might be a bit confusing. How about something like `invalid name: cannot have empty name`

File: internal/app/singularity/remote_add.go [hunk 4]
```diff
@@ -10,12 +10,21 @@ import (
...
+	}
+	if strings.TrimSpace(uri) == "" {
+		return fmt.Errorf("invalid URI: %s", uri)
```
Comment (ikaneshiro) [inline internal/app/singularity/remote_add.go]: Same here

**Review by ikaneshiro — CHANGES_REQUESTED** [review]

File: internal/app/singularity/remote_add.go (hunk 3 above)
Comment (gvallee) [reply internal/app/singularity/remote_add.go]: Good point, fixed

File: internal/app/singularity/remote_add.go (hunk 4 above)
Comment (gvallee) [reply internal/app/singularity/remote_add.go]: Done

File: internal/app/singularity/remote_add_test.go (hunk 2 above)
Comment (gvallee) [reply internal/app/singularity/remote_add_test.go]: Very sorry about that. I am getting tired and tried to go too fast. It is now fixed.

**Comment by gvallee** [comment]
@ikaneshiro I would like now to have the confirmation that cases 12, 13, 29 and 30 are indeed supposed to succeed. My impression is that they should not but I am absolutely not sure.

**Comment by ikaneshiro** [comment]
@gvallee I think those should pass.

Actually I think we can remove the tests in add that are only testing for `invalid remote name` or `invalid URI` since users should be allowed to add remotes with whatever names and locations they want. The only case to test for name and uri would be an empty entry.

Testing for invalid names makes sense for the remove command though since the only valid names are the ones currently configured

**Comment by gvallee** [comment]
@ikaneshiro It makes sense. I was concerned about the URI since some cases clearly do not use a valid URI but I understand your point. I will update the code ASAP and maybe then we could merge that PR; I think this work clarify a few things. Thanks again for your time!

**Comment by gvallee** [comment]
All done

**Review by ikaneshiro — APPROVED** [review]

**Pull request merged**